### ISSUE TITLE
For idea plugin, `.add`ing an excluded folder does not work, only reassignment works

### ISSUE THREAD

**Issue by matklad** [gradle/gradle#8749]
### Expected Behavior

For

```
idea {
    module {
        excludeDirs.add(file("testData"))
    }
}
```

I expect that `testData` will be an excluded folder in IDE.

### Current Behavior

`testData` is not an excluded folder. However, if I use more verbose `excludeDirs = excludeDirs + file("testData")`, `testData` is indeed excluded. 

### Context

This seems like a bug either with DSL or `IdeaModule`. Note that `generatedSourceDirs.add` works fine. Here's the project where I am facing the issue: https://github.com/intellij-rust/intellij-rust/blob/c150f089e59a99e61ddbd16a178d096b17116010/build.gradle.kts

**Comment by big-guy** [comment]
This is a quirk of the way the default values for the excluded directories works.  The ultimate fix for this is to stop using a plain `List` and use a `ListProperty`.

`excludeDirs.add(file("testData"))` adds the new directory to a List that's immediately thrown away.

The workaround is to use `+=`, which is equivalent to the verbose option mentioned above: `excludeDirs += file("testData")`

We're not going to get to fixing this any time soon, but we would eventually get to this as we Property-ize more tasks/APIs.

**Comment by Vampire** [comment]
`testSourceDirs += allJava.sourceDirectories.files` does not work with Kotlin DSL due to
```
Assignment operators ambiguity: 
public operator fun <T> Set<File!>.plus(elements: Iterable<File!>): Set<File!> defined in kotlin.collections
@InlineOnly public inline operator fun <T> MutableCollection<in File!>.plusAssign(elements: Iterable<File!>): Unit defined in kotlin.collections
```

Neither does `testSourceDirs.plusAssign(allJava.sourceDirectories.files)` which just has no visual effect.

You have to actually use `testSourceDirs = testSourceDirs + allJava.sourceDirectories.files` to make it work properly.

**Comment by oehme** [comment]
This makes the Kotlin DSL look really silly. This could be fixed pretty easily if the Set was initialized outside of the closure and then the same instance would be returned every time: https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java#L302

**Comment by Vampire** [comment]
Will you create a PR with that change? :-)

**Comment by oehme** [comment]
If the Gradle team agrees, yeah I can do that.

**Comment by big-guy** [comment]
@oehme that sounds reasonable to me

**Comment by Vampire** [comment]
I was under the impression I had seen this fixed, but I right now cannot reproduce the fixed behavior neither with Gradle DSL nor Kotlin DSL.
I still get the faulty behavior. :-(

**Comment by oehme** [comment]
Did you try with a basic project to make sure nothing else is re-assigning the field?

**Comment by Vampire** [comment]
Yes

**Comment by oehme** [comment]
I checked and the [integration with the Java plugin](https://github.com/gradle/gradle/blob/master/subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/IdeaPlugin.java#L376) is causing that issue. That wasn't covered by my unit test, which only uses the plain IDEA plugin.

**Comment by Vampire** [comment]
Prepared a PR for it at https://github.com/gradle/gradle/pull/17993

**Comment by ljacomet** [comment]
Opening back since there is a PR looking at fixing this.

**Comment by ljacomet** [comment]
The full fix for this will be in Gradle 7.3 but a partial fix was already available in Gradle 7.1
